compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64LIRGenerator.java

```diff
@@ -384,31 +385,44 @@ protected boolean emitCompare(PlatformKind cmpKind, Value a, Value b, Condition
                 bExt = arithmeticLIRGen.emitSignExtend(b, compareBytes * 8, 64);
             }
 
-            if (LIRValueUtil.isVariable(bExt)) {
-                left = load(bExt);
+            /*
+             * The AArch64 comparison instruction can treat register 31 as the stack pointer
+             * register for the left operand, but not for the right operand.
+             */
+            boolean aIsStackPointer = ValueUtil.isRegister(aExt) && ValueUtil.asRegister(aExt).equals(AArch64.sp);
+            boolean bIsStackPointer = ValueUtil.isRegister(bExt) && ValueUtil.asRegister(bExt).equals(AArch64.sp);
+
+            if (aIsStackPointer && bIsStackPointer) {
+                // move b to a non-sp register
+                bIsStackPointer = false;
+                bExt = emitMove(bExt);
+            }
+
+            if (bIsStackPointer || (!aIsStackPointer && LIRValueUtil.isVariable(bExt))) {
+                left = bExt;
                 right = loadNonConst(aExt);
                 mirrored = true;
             } else {
-                left = load(aExt);
+                left = aExt;
                 right = loadNonConst(bExt);
                 mirrored = false;
             }
-            append(new AArch64Compare.CompareOp(left, loadNonCompareConst(right)));
+            append(new AArch64Compare.CompareOp(loadReg(left), loadNonCompareConst(right)));
```

Comment:
All the changes in this patch make sense individually. But, I can't see what's actually removing the unnecessary loads. Is it the use of loadReg instead or load?

Reply:
Yes. Using load will introduce moves for values already assigned a registervalue. In general it would be a good idea to remove all unneeded `load`s in the code